### DISCUSSION TITLE
Unable to find Prisma Client in GraphQL context. Please provide it under the `context.prisma` key.

### DISCUSSION THREAD

**Discussion by omar-dulaimi** [MichalLytek/typegraphql-prisma#111] [Q&A] [answered]
Hello,

After Prisma confirmed that RDS Proxy would be useless because of pinning, this means that we can't deploy live. So a Prisma member suggested that we create a client when a request comes and disconnect once it's done. So that's basically what I want to do.

I defined a parameter decorator to generate a PrismaInstance:

```
export function PrismaInstance(): ParameterDecorator {
    return createParamDecorator(() => {
        return new PrismaClient({
            log: ['error', 'info', 'query', 'warn'],
            datasources: {
                db: {
                    url: databaseUrl
                },
            },
        });
    });
}

```

And in each resolver I do this:

```
    @TypeGraphQL.Query(_returns => [Car], {
        nullable: false,
    })
    async cars(
        @TypeGraphQL.Args() args: FindManyCarArgs,
        @PrismaInstance() prisma: PrismaClient,
    ): Promise<Car[]> {
        const response = await prisma.car.findMany(args);
        await prisma.$disconnect();
        return response;
    }
```

So I have two questions:

1- Is there a better way to achieve the above?

2- Since I'm defining cusotm resolvers that rely on the generated resolvers, how can I fix the error coming from this line

`getPrismaFromContext(ctx)`

The problem is that I don't add the prisma client on the apollo server anymore since I switched to using the parameter decorator above.

```
const server = new ApolloServer({
    schema,
    context: ({ event }) => ({
        // ...prismaContext,
        headers: event.headers,
    }),
    validationRules,
    formatError,
    introspection: true,
    playground: {
        endpoint: process.env.PLAYGROUND_ENDPOINT,
    },
});
```

**Comment by MichalLytek** [comment, marked as answer]
Parameter decorator is not a right solution. It will connect and disconnect for every field resolver, meaning prisma batching won't work. And will be slow as turle 😅 

Instead you should initialize prisma client in the `context` helper, so single graphql request which is handled by separate lambda will use single Prisma Client instance.

However you may need to dig in into Apollo Server docs to find a way to use hooks and plugins in order to disconnect from prisma, as TypeGraphQL is not responsible for the whole GraphQL execution pipeline, it only creates resolver handlers for GraphQL Schema.

**Reply by omar-dulaimi** [reply]
Thank you for pointing that out. I'll try that and hopefully it will work.

**Reply by omar-dulaimi** [reply]
So this is what I have now:
```
const server = new ApolloServer({
    schema,
    context: ({ event }) => ({
        prisma: new PrismaClient({
            datasources: {
                db: {
                    url: databaseUrl
                },
            },
        }),
        headers: event.headers,
    }),
    validationRules,
    formatError,
    introspection: true,
    playground: {
        endpoint: process.env.PLAYGROUND_ENDPOINT,
    },
    plugins: [
        {
            requestDidStart() {
                return {
                    async willSendResponse(requestContext) {
                        await requestContext.context.prisma.$disconnect();
                    },
                };
            },
        },
    ],
});
```

Do you have any feedback?

I tried running a load test with k6 using this config:

```
export let options = {
  vus: 100, // virtual users
  duration: "3s",
  httpDebug: 'full',
};
```

But I get a lot of "too many connections" errors.

**Reply by MichalLytek** [reply]
Log prisma creation and disconnect so see if it really works and disconects.
Then you can go back to Prisma support as now the solution is not related to TypeGraphQL.

**Reply by omar-dulaimi** [reply]
It does create and disconnect. Waiting for Prisma response now. thank you for your help.
